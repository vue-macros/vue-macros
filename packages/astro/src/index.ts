import { resolveOptions, type Options } from 'vue-macros'
import VueMacros from 'vue-macros/vite'
import type { AstroIntegration, ViteUserConfig } from 'astro'
import type { Plugin } from 'vite'

function findPluginAndRemove(
  name: string,
  plugins: ViteUserConfig['plugins'],
): Plugin | undefined {
  const idx = plugins!.findIndex(
    (plugin) => plugin && 'name' in plugin && plugin.name === name,
  )
  if (idx === -1) return
  const plugin = plugins![idx]
  plugins!.splice(idx, 1)
  return plugin as any
}

export default function (options?: Options): AstroIntegration {
  return {
    name: '@vue-macros/astro',
    hooks: {
      'astro:config:setup': async ({ config }) => {
        const resolvedOptions = await resolveOptions(options || {})
        const vue = findPluginAndRemove('vite:vue', config.vite.plugins)
        const vueJsx = findPluginAndRemove('vite:vue-jsx', config.vite.plugins)

        const vueMacrosPlugins = await VueMacros({
          ...resolvedOptions,
          plugins: {
            vue,
            vueJsx,
          },
        })
        config.vite.plugins ||= []
        config.vite.plugins.push(...vueMacrosPlugins)
      },
    },
  }
}
