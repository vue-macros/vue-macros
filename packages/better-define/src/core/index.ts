import {
  analyzeSFC,
  genRuntimePropDefinition,
  type ResultAsync,
  type TSEmits,
  type TSProps,
} from '@vue-macros/api'
import {
  DEFINE_EMITS,
  escapeKey,
  generateTransform,
  importHelperFn,
  MagicStringAST,
  parseSFC,
  type CodeTransform,
} from '@vue-macros/common'
import { ok, safeTry } from 'neverthrow'

export function transformBetterDefine(
  code: string,
  id: string,
  isProduction = false,
): ResultAsync<CodeTransform | void> {
  return safeTry(async function* () {
    const s = new MagicStringAST(code)
    const sfc = parseSFC(code, id)
    if (!sfc.scriptSetup) return ok()

    const offset = sfc.scriptSetup.loc.start.offset
    const result = yield* analyzeSFC(s, sfc)
    if (result.props) {
      yield* processProps(result.props)
    }
    if (result.emits) {
      processEmits(result.emits)
    }

    return ok(generateTransform(s, id))

    function processProps(props: TSProps) {
      return safeTry(async function* () {
        const runtimeDefs = yield* props.getRuntimeDefinitions()

        const runtimeDecls = `{\n  ${Object.entries(runtimeDefs)
          .map(([key, { type, required, default: defaultDecl }]) => {
            let defaultString = ''
            if (defaultDecl) defaultString = defaultDecl('default')

            const properties: string[] = []
            if (!isProduction) properties.push(`required: ${required}`)
            if (defaultString) properties.push(defaultString)

            return `${escapeKey(key)}: ${genRuntimePropDefinition(
              type,
              isProduction,
              properties,
            )}`
          })
          .join(',\n  ')}\n}`

        let decl = runtimeDecls
        if (props.withDefaultsAst && !props.defaults) {
          // dynamic defaults
          decl = `${importHelperFn(
            s,
            offset,
            'mergeDefaults',
          )}(${decl}, ${s.sliceNode(props.withDefaultsAst.arguments[1], {
            offset,
          })})`
          // add helper
        }
        decl = `defineProps(${decl})`

        s.overwriteNode(props.withDefaultsAst || props.definePropsAst, decl, {
          offset,
        })

        return ok()
      })
    }

    function processEmits(emits: TSEmits) {
      const runtimeDecls = `[${Object.keys(emits.definitions)
        .map((name) => JSON.stringify(name))
        .join(', ')}]`
      s.overwriteNode(
        emits.defineEmitsAst,
        `${DEFINE_EMITS}(${runtimeDecls})`,
        { offset },
      )
    }
  })
}
