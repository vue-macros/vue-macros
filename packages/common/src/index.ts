export * from 'magic-string-ast'
export * from 'ast-kit'
export {
  createFilter as createRollupFilter,
  normalizePath,
  type FilterPattern,
} from 'unplugin-utils'

export * from './ast'
export * from './constants'
export * from './dep'
export * from './options'
export * from './types'
export * from './unplugin'
export * from './vue'
