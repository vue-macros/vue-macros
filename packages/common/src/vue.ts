import {
  parse,
  type SFCDescriptor,
  type SFCParseResult,
  type SFCScriptBlock as SFCScriptBlockMixed,
} from '@vue/compiler-sfc'
import { babelParse, getLang, resolveString } from 'ast-kit'
import { REGEX_VUE_SFC } from './constants'
import type { Node, Program } from '@babel/types'
import type { MagicString, MagicStringAST } from 'magic-string-ast'

export type SFCScriptBlock = Omit<
  SFCScriptBlockMixed,
  'scriptAst' | 'scriptSetupAst'
>

export type SFC = Omit<SFCDescriptor, 'script' | 'scriptSetup'> & {
  sfc: SFCParseResult
  script?: SFCScriptBlock | null
  scriptSetup?: SFCScriptBlock | null
  lang: string | undefined
  getScriptAst: () => Program | undefined
  getSetupAst: () => Program | undefined
  offset: number
} & Pick<SFCParseResult, 'errors'>

export function parseSFC(code: string, id: string): SFC {
  const sfc = parse(code, {
    filename: id,
  })
  const { descriptor, errors } = sfc

  const scriptLang = sfc.descriptor.script?.lang
  const scriptSetupLang = sfc.descriptor.scriptSetup?.lang

  if (
    sfc.descriptor.script &&
    sfc.descriptor.scriptSetup &&
    (scriptLang || 'js') !== (scriptSetupLang || 'js')
  ) {
    throw new Error(
      `[vue-macros] <script> and <script setup> must have the same language type.`,
    )
  }

  const lang = scriptLang || scriptSetupLang

  return Object.assign({}, descriptor, {
    sfc,
    lang,
    errors,
    offset: descriptor.scriptSetup?.loc.start.offset ?? 0,
    getSetupAst() {
      if (!descriptor.scriptSetup) return
      return babelParse(descriptor.scriptSetup.content, lang, {
        plugins: [['importAttributes', { deprecatedAssertSyntax: true }]],
        cache: true,
      })
    },
    getScriptAst() {
      if (!descriptor.script) return
      return babelParse(descriptor.script.content, lang, {
        plugins: [['importAttributes', { deprecatedAssertSyntax: true }]],
        cache: true,
      })
    },
  } satisfies Partial<SFC>)
}

export function getFileCodeAndLang(
  code: string,
  id: string,
): { code: string; lang: string } {
  if (!REGEX_VUE_SFC.test(id)) {
    return {
      code,
      lang: getLang(id),
    }
  }

  const sfc = parseSFC(code, id)
  const scriptCode = sfc.script?.content ?? ''
  return {
    code: sfc.scriptSetup
      ? `${scriptCode}\n;\n${sfc.scriptSetup.content}`
      : scriptCode,
    lang: sfc.lang ?? 'js',
  }
}

export function addNormalScript({ script, lang }: SFC, s: MagicString) {
  return {
    start(): number {
      if (script) return script.loc.end.offset

      const attrs = lang ? ` lang="${lang}"` : ''
      s.prependLeft(0, `<script${attrs}>`)
      return 0
    },

    end(): void {
      if (!script) s.appendRight(0, `\n</script>\n`)
    },
  }
}

export function removeMacroImport(
  node: Node,
  s: MagicStringAST,
  offset: number,
): true | undefined {
  if (
    node.type === 'ImportDeclaration' &&
    node.attributes?.some(
      (attr) =>
        resolveString(attr.key) === 'type' && attr.value.value === 'macro',
    )
  ) {
    s.removeNode(node, { offset })
    return true
  }
}
