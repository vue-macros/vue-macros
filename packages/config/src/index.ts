export * from './config'
export * from './define'
export * from './options'
