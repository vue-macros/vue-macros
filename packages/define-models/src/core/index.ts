import {
  DEFINE_EMITS,
  DEFINE_MODELS,
  DEFINE_MODELS_DOLLAR,
  DEFINE_PROPS,
  generateTransform,
  HELPER_PREFIX,
  importHelperFn,
  isCallOf,
  MagicStringAST,
  parseSFC,
  REPO_ISSUE_URL,
  resolveObjectKey,
  WITH_DEFAULTS,
  type CodeTransform,
} from '@vue-macros/common'
import { extractIdentifiers, walkAST } from 'ast-walker-scope'
import { emitHelperId, useVmodelHelperId } from './helper'
import type {
  Identifier,
  LVal,
  Node,
  ObjectPattern,
  TSInterfaceBody,
  TSType,
  TSTypeLiteral,
  VariableDeclaration,
} from '@babel/types'

export function transformDefineModels(
  code: string,
  id: string,
): CodeTransform | undefined {
  let hasDefineProps = false
  let hasDefineEmits = false
  let hasDefineModels = false

  let propsTypeDecl: TSType | undefined
  let propsDestructureDecl: Node | undefined
  let emitsTypeDecl: TSType | undefined
  let emitsIdentifier: string | undefined

  let runtimeDefineFn: string | undefined

  let modelDecl: Node | undefined
  let modelDeclKind: string | undefined
  let modelTypeDecl: TSType | undefined
  let modelIdentifier: string | undefined
  let modelDestructureDecl: ObjectPattern | undefined

  const modelIdentifiers = new Set<Identifier>()
  let mode: 'reactivity-transform' | 'runtime' | undefined

  function processDefinePropsOrEmits(node: Node, declId?: LVal) {
    if (isCallOf(node, WITH_DEFAULTS)) {
      node = node.arguments[0]
    }

    let type: 'props' | 'emits'
    if (isCallOf(node, DEFINE_PROPS)) {
      type = 'props'
    } else if (isCallOf(node, DEFINE_EMITS)) {
      type = 'emits'
    } else {
      return false
    }
    const fnName = type === 'props' ? DEFINE_PROPS : DEFINE_EMITS

    if (node.arguments[0]) {
      runtimeDefineFn = fnName
      return false
    }

    if (type === 'props') hasDefineProps = true
    else hasDefineEmits = true

    const typeDecl = node.typeParameters?.params?.[0]
    if (!typeDecl)
      throw new SyntaxError(
        `${fnName}() expected a type parameter when used with ${DEFINE_MODELS}.`,
      )

    if (type === 'props') propsTypeDecl = typeDecl
    else emitsTypeDecl = typeDecl

    if (declId) {
      if (type === 'props' && declId.type === 'ObjectPattern') {
        propsDestructureDecl = declId
      } else if (type === 'emits' && declId.type === 'Identifier') {
        emitsIdentifier = declId.name
      }
    } else if (type === 'emits') {
      emitsIdentifier = `_${DEFINE_MODELS}_emit`
      s.prependRight(setupOffset + node.start!, `const ${emitsIdentifier} = `)
    }

    return true
  }

  function processDefineModels(
    node: Node,
    declId?: LVal,
    kind?: VariableDeclaration['kind'],
  ) {
    if (isCallOf(node, DEFINE_MODELS)) mode = 'runtime'
    else if (isCallOf(node, DEFINE_MODELS_DOLLAR)) mode = 'reactivity-transform'
    else return false

    if (hasDefineModels) {
      throw new SyntaxError(`duplicate ${DEFINE_MODELS}() call`)
    }
    hasDefineModels = true
    modelDecl = node

    modelTypeDecl = node.typeParameters?.params[0]
    if (!modelTypeDecl) {
      throw new SyntaxError(`expected a type parameter for ${DEFINE_MODELS}.`)
    }

    if (mode === 'reactivity-transform' && declId) {
      const ids = extractIdentifiers(declId)
      ids.forEach((id) => modelIdentifiers.add(id))

      if (declId.type === 'ObjectPattern') {
        modelDestructureDecl = declId
        for (const property of declId.properties) {
          if (property.type === 'RestElement') {
            throw new SyntaxError('rest element is not supported')
          }
        }
      } else {
        modelIdentifier = scriptSetup!.loc.source.slice(
          declId.start!,
          declId.end!,
        )
      }
    }
    if (kind) modelDeclKind = kind

    return true
  }

  function extractPropsDefinitions(node: TSTypeLiteral | TSInterfaceBody) {
    const members = node.type === 'TSTypeLiteral' ? node.members : node.body
    const map: Record<
      string,
      {
        typeAnnotation: string
        options?: Record<string, string>
      }
    > = Object.create(null)

    for (const m of members) {
      if (
        (m.type === 'TSPropertySignature' || m.type === 'TSMethodSignature') &&
        m.key.type === 'Identifier'
      ) {
        const type = m.typeAnnotation?.typeAnnotation
        let typeAnnotation = ''
        let options: Record<string, string> | undefined
        if (type) {
          typeAnnotation += `${m.optional ? '?' : ''}: `
          if (
            type.type === 'TSTypeReference' &&
            type.typeName.type === 'Identifier' &&
            type.typeName.name === 'ModelOptions' &&
            type.typeParameters?.type === 'TSTypeParameterInstantiation' &&
            type.typeParameters.params[0]
          ) {
            typeAnnotation += setupContent.slice(
              type.typeParameters.params[0].start!,
              type.typeParameters.params[0].end!,
            )
            if (type.typeParameters.params[1]?.type === 'TSTypeLiteral') {
              options = Object.create(null)
              for (const m of type.typeParameters.params[1].members) {
                if (
                  (m.type === 'TSPropertySignature' ||
                    m.type === 'TSMethodSignature') &&
                  m.key.type === 'Identifier'
                ) {
                  const type = m.typeAnnotation?.typeAnnotation
                  if (type)
                    options![setupContent.slice(m.key.start!, m.key.end!)] =
                      setupContent.slice(type.start!, type.end!)
                }
              }
            }
          } else typeAnnotation += setupContent.slice(type.start!, type.end!)
        }

        map[m.key.name] = { typeAnnotation, options }
      }
    }
    return map
  }

  function rewriteMacros() {
    rewriteDefines()
    if (mode === 'runtime') {
      rewriteRuntime()
    }

    function rewriteDefines() {
      const propsText = Object.entries(map)
        .map(
          ([key, { typeAnnotation }]) => `${getPropKey(key)}${typeAnnotation}`,
        )
        .join(';\n')

      const emitsText = Object.entries(map)
        .map(
          ([key, { typeAnnotation }]) =>
            `(evt: '${getEventKey(key)}', value${typeAnnotation}): void;`,
        )
        .join('\n  ')

      if (hasDefineProps) {
        s.overwriteNode(
          propsTypeDecl!,
          `(${s.sliceNode(propsTypeDecl!, {
            offset: setupOffset,
          })}) & {\n  ${propsText}\n}`,
          { offset: setupOffset },
        )
        if (
          mode === 'reactivity-transform' &&
          propsDestructureDecl &&
          modelDestructureDecl
        )
          for (const property of modelDestructureDecl.properties) {
            const text = code.slice(
              setupOffset + property.start!,
              setupOffset + property.end!,
            )
            s.appendLeft(
              setupOffset + propsDestructureDecl.start! + 1,
              `${text}, `,
            )
          }
      } else {
        let text = ''
        const kind = modelDeclKind || 'let'
        if (mode === 'reactivity-transform') {
          if (modelIdentifier) {
            text = modelIdentifier
          } else if (modelDestructureDecl) {
            text = code.slice(
              setupOffset + modelDestructureDecl.start!,
              setupOffset + modelDestructureDecl.end!,
            )
          }
        }

        s.appendRight(
          setupOffset,
          `\n${text ? `${kind} ${text} = ` : ''}defineProps<{
  ${propsText}
}>();`,
        )
      }

      if (hasDefineEmits) {
        s.overwriteNode(
          emitsTypeDecl!,
          `(${s.sliceNode(emitsTypeDecl!, {
            offset: setupOffset,
          })}) & {\n  ${emitsText}\n}`,
          { offset: setupOffset },
        )
      } else {
        emitsIdentifier = `${HELPER_PREFIX}emit`
        s.appendRight(
          setupOffset,
          `\n${
            mode === 'reactivity-transform' ? `const ${emitsIdentifier} = ` : ''
          }defineEmits<{
  ${emitsText}
}>();`,
        )
      }
    }
  }

  function rewriteRuntime() {
    const text = `${importHelperFn(
      s,
      setupOffset,
      'useVModel',
      useVmodelHelperId,
      true,
    )}(${Object.entries(map)
      .map(([name, { options }]) => {
        const prop = getPropKey(name, true)
        const evt = getEventKey(name, true)
        if (!prop && !evt && !options) return stringifyValue(name)

        const args = [name, prop, evt].map((arg) => stringifyValue(arg))
        if (options) {
          const str = Object.entries(options)
            .map(([k, v]) => `  ${stringifyValue(k)}: ${v}`)
            .join(',\n')
          args.push(`{\n${str}\n}`)
        }

        return `[${args.join(', ')}]`
      })
      .join(', ')})`
    s.overwriteNode(modelDecl!, text, { offset: setupOffset })
  }

  function processAssignModelVariable() {
    if (!emitsIdentifier)
      throw new Error(
        `Identifier of returning value of ${DEFINE_EMITS} is not found, please report this issue.\n${REPO_ISSUE_URL}`,
      )

    function overwrite(
      node: Node,
      id: Identifier,
      value: string,
      original = false,
    ) {
      const eventName = aliasMap[id.name]
      const content = `${importHelperFn(
        s,
        setupOffset,
        'emitHelper',
        emitHelperId,
        true,
      )}(${emitsIdentifier}, '${getEventKey(String(eventName))}', ${value}${
        original ? `, ${id.name}` : ''
      })`
      s.overwriteNode(node, content, { offset: setupOffset })
    }

    walkAST(setupAst!, {
      leave(node) {
        if (node.type === 'AssignmentExpression') {
          if (node.left.type !== 'Identifier') return
          const id = this.scope[node.left.name] as Identifier
          if (!modelIdentifiers.has(id)) return

          const left = s.sliceNode(node.left, { offset: setupOffset })
          let right = s.sliceNode(node.right, { offset: setupOffset })
          if (node.operator !== '=') {
            right = `${left} ${node.operator.replace(/=$/, '')} ${right}`
          }

          overwrite(node, id, right)
        } else if (node.type === 'UpdateExpression') {
          if (node.argument.type !== 'Identifier') return
          const id = this.scope[node.argument.name] as Identifier
          if (!modelIdentifiers.has(id)) return

          let value = node.argument.name
          if (node.operator === '++') value += ' + 1'
          else value += ' - 1'

          overwrite(node, id, value, !node.prefix)
        }
      },
    })
  }

  if (!code.includes(DEFINE_MODELS)) return
  const { scriptSetup, getSetupAst } = parseSFC(code, id)
  if (!scriptSetup) return

  const setupOffset = scriptSetup.loc.start.offset
  const setupContent = scriptSetup.content
  const setupAst = getSetupAst()!.body

  const s = new MagicStringAST(code)

  // process <script setup>
  for (const node of setupAst) {
    if (node.type === 'ExpressionStatement') {
      processDefinePropsOrEmits(node.expression)

      if (
        processDefineModels(node.expression) &&
        mode === 'reactivity-transform'
      )
        s.remove(node.start! + setupOffset, node.end! + setupOffset)
    } else if (node.type === 'VariableDeclaration' && !node.declare) {
      const total = node.declarations.length
      let left = total

      for (let i = 0; i < total; i++) {
        const decl = node.declarations[i]
        if (decl.init) {
          processDefinePropsOrEmits(decl.init, decl.id)

          if (
            processDefineModels(decl.init, decl.id, node.kind) &&
            mode === 'reactivity-transform'
          ) {
            if (left === 1) {
              s.remove(node.start! + setupOffset, node.end! + setupOffset)
            } else {
              let start = decl.start! + setupOffset
              let end = decl.end! + setupOffset
              if (i < total - 1) {
                // not the last one, locate the start of the next
                end = node.declarations[i + 1].start! + setupOffset
              } else {
                // last one, locate the end of the prev
                start = node.declarations[i - 1].end! + setupOffset
              }
              s.remove(start, end)
              left--
            }
          }
        }
      }
    }
  }

  if (!modelTypeDecl) return

  if (runtimeDefineFn)
    throw new SyntaxError(
      `${runtimeDefineFn}() cannot accept non-type arguments when used with ${DEFINE_MODELS}()`,
    )

  if (modelTypeDecl.type !== 'TSTypeLiteral') {
    throw new SyntaxError(
      `type argument passed to ${DEFINE_MODELS}() must be a literal type, or a reference to an interface or literal type.`,
    )
  }

  const map = extractPropsDefinitions(modelTypeDecl)
  const aliasMap: Record<string, string | number> = Object.create(null)
  if (modelDestructureDecl)
    for (const p of modelDestructureDecl.properties) {
      if (p.type !== 'ObjectProperty') continue
      try {
        const key = resolveObjectKey(p)
        if (p.value.type !== 'Identifier') continue
        aliasMap[p.value.name] = key
      } catch {}
    }

  // const defaults = resolveObjectExpression(defaultsAst)
  // if (!defaults) return { defaultsAst }
  rewriteMacros()

  if (mode === 'reactivity-transform' && hasDefineModels)
    processAssignModelVariable()

  return generateTransform(s, id)
}

function stringifyValue(value: string | undefined) {
  return value !== undefined ? JSON.stringify(value) : 'undefined'
}

function getPropKey(key: string, omitDefault = false) {
  return !omitDefault ? key : undefined
}

function getEventKey(key: string, omitDefault = false) {
  return !omitDefault ? `update:${key}` : undefined
}
