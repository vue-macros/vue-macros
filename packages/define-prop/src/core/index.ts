import { inferRuntimeType, resolveTSReferencedType } from '@vue-macros/api'
import {
  DEFINE_PROP,
  DEFINE_PROP_DOLLAR,
  DEFINE_PROPS,
  generateTransform,
  HELPER_PREFIX,
  importHelperFn,
  isCallOf,
  MagicStringAST,
  parseSFC,
  walkAST,
  type CodeTransform,
} from '@vue-macros/common'
import { helperId } from './helper'
import { johnsonEdition } from './johnson-edition'
import { kevinEdition } from './kevin-edition'
import type * as t from '@babel/types'

export * from './utils'
export * from './kevin-edition'
export * from './johnson-edition'

export const PROPS_VARIABLE_NAME: '__MACROS_props' = `${HELPER_PREFIX}props`
export type Edition = 'kevinEdition' | 'johnsonEdition'

export async function transformDefineProp(
  code: string,
  id: string,
  edition: Edition = 'kevinEdition',
  isProduction = false,
): Promise<CodeTransform | undefined> {
  if (!code.includes(DEFINE_PROP)) return

  const sfc = parseSFC(code, id)
  if (!sfc.scriptSetup) return

  const setupAst = sfc.getSetupAst()!

  const offset = sfc.scriptSetup.loc.start.offset
  const s = new MagicStringAST(code)

  const { walkCall, genRuntimeProps } = (
    edition === 'kevinEdition' ? kevinEdition : johnsonEdition
  )({ s, offset, resolveTSType })

  let definePropsCall: t.CallExpression | undefined
  const parentMap = new WeakMap<t.Node, t.Node | undefined | null>()
  walkAST<t.Node>(setupAst, {
    enter: (node, parent) => {
      if (isCallOf(node, DEFINE_PROPS)) {
        definePropsCall = node
      }

      parentMap.set(node, parent)

      if (!isCallOf(node, [DEFINE_PROP, DEFINE_PROP_DOLLAR])) return

      const isCallOfDollar =
        isCallOf(parent, '$') && isCallOf(node, DEFINE_PROP)

      const isReactiveTransform =
        isCallOfDollar || isCallOf(node, DEFINE_PROP_DOLLAR)

      const propName = walkCall(
        node,
        isCallOfDollar ? parentMap.get(parent) : parent,
      )
      s.overwriteNode(
        isCallOfDollar ? parent : node,
        `${isReactiveTransform ? '$(' : ''}${importHelperFn(
          s,
          offset,
          'toRef',
        )}(__props, ${JSON.stringify(propName)})${
          isReactiveTransform ? ')' : ''
        }`,
        { offset },
      )
    },
  })

  const runtimeProps = await genRuntimeProps(isProduction)
  if (!runtimeProps) return

  if (definePropsCall?.typeParameters) {
    throw new SyntaxError(
      `defineProp cannot be used with defineProps<T>() in the same component.`,
    )
  }

  if (definePropsCall && definePropsCall.arguments.length > 0) {
    const originalProps = s.sliceNode(definePropsCall.arguments[0], { offset })
    const normalizePropsOrEmits = importHelperFn(
      s,
      offset,
      'default',
      'normalizePropsOrEmits',
      helperId,
    )
    s.overwriteNode(
      definePropsCall.arguments[0],
      `{ ...${normalizePropsOrEmits}(${originalProps}), ...${normalizePropsOrEmits}(${runtimeProps}) }`,
      {
        offset,
      },
    )
  } else {
    s.prependLeft(
      offset,
      `\nconst ${PROPS_VARIABLE_NAME} = defineProps(${runtimeProps});\n`,
    )
  }

  return generateTransform(s, id)

  function resolveTSType(type: t.TSType): Promise<string[] | undefined> {
    return resolveTSReferencedType({
      scope: {
        kind: 'file',
        filePath: id,
        content: sfc.scriptSetup!.content,
        ast: setupAst.body,
      },
      type,
    })
      .map(
        (resolved) =>
          resolved && inferRuntimeType(resolved).unwrapOr(undefined),
      )
      .unwrapOr(undefined)
  }
}
