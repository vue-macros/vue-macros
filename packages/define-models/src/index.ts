import {
  createFilter,
  detectVueVersion,
  FilterFileType,
  getFilterPattern,
  normalizePath,
  type BaseOptions,
  type MarkRequired,
} from '@vue-macros/common'
import { generatePluginName } from '#macros' with { type: 'macro' }
import {
  createUnplugin,
  type UnpluginContextMeta,
  type UnpluginInstance,
} from 'unplugin'
import { transformDefineModels } from './core'
import {
  emitHelperCode,
  emitHelperId,
  helperPrefix,
  useVmodelHelperCode,
  useVmodelHelperId,
} from './core/helper'

export interface Options extends BaseOptions {}

export type OptionsResolved = MarkRequired<Options, 'include' | 'version'>

function resolveOptions(
  options: Options,
  framework: UnpluginContextMeta['framework'],
): OptionsResolved {
  const version = options.version || detectVueVersion()
  const include = getFilterPattern(
    [FilterFileType.VUE_SFC_WITH_SETUP, FilterFileType.SETUP_SFC],
    framework,
  )
  return {
    include,
    ...options,
    version,
  }
}

const name = generatePluginName()

const plugin: UnpluginInstance<Options | undefined, false> = createUnplugin(
  (userOptions = {}, { framework }) => {
    const options = resolveOptions(userOptions, framework)
    const filter = createFilter(options)

    return {
      name,
      enforce: 'pre',

      resolveId(id) {
        if (normalizePath(id).startsWith(helperPrefix)) return id
      },

      loadInclude(id) {
        return normalizePath(id).startsWith(helperPrefix)
      },

      load(_id) {
        const id = normalizePath(_id)
        if (id === emitHelperId) return emitHelperCode
        else if (id === useVmodelHelperId) return useVmodelHelperCode
      },

      transformInclude: filter,
      transform(code, id) {
        return transformDefineModels(code, id)
      },
    }
  },
)
export default plugin
