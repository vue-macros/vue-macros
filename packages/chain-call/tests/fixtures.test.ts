import { resolve } from 'node:path'
import {
  rollupBuild,
  RollupRemoveVueFilePathPlugin,
  RollupVue,
  testFixtures,
  UnpluginOxc,
} from '@vue-macros/test-utils'
import { describe } from 'vitest'
import { transformChainCall } from '../src/core'
import VueChainCall from '../src/rollup'

describe('fixtures', async () => {
  await testFixtures(
    ['tests/fixtures/*.vue', '!tests/fixtures/definePropsRefs.vue'],
    (_, id) =>
      rollupBuild(id, [
        VueChainCall(),
        RollupVue(),
        RollupRemoveVueFilePathPlugin(),
        UnpluginOxc.rollup(),
      ]),
    {
      cwd: resolve(__dirname, '..'),
      promise: true,
    },
  )
  await testFixtures(
    import.meta.glob<string>('./fixtures/definePropsRefs.vue', {
      eager: true,
      query: '?raw',
      import: 'default',
    }),
    (_, id, code) => transformChainCall(code, id)?.code,
  )
})
