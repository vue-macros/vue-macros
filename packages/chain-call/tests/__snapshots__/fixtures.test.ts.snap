// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`fixtures > tests/fixtures/basic.vue 1`] = `
"// basic.js
import { defineComponent } from 'vue';
import { expectTypeOf } from 'expect-type';
import _export_sfc from '[NULL]/plugin-vue/export-helper';

var _sfc_main = /* @__PURE__ */ defineComponent({
	__name: "basic",
	props: {
		foo: {
			type: String,
			required: false,
			default: "111"
		},
		bar: {
			type: Number,
			required: false,
			default: 111
		},
		baz: {
			type: Boolean,
			required: false,
			default: false
		}
	},
	setup(__props, { expose: __expose }) {
		__expose();
		const props = __props;
		expectTypeOf(props).toEqualTypeOf();
		const __returned__ = {
			props,
			get expectTypeOf() {
				return expectTypeOf;
			}
		};
		Object.defineProperty(__returned__, "__isScriptSetup", {
			enumerable: false,
			value: true
		});
		return __returned__;
	}
});

var basic = /* @__PURE__ */ _export_sfc(_sfc_main, [__FILE__]);

export { basic as default };
"
`;

exports[`fixtures > tests/fixtures/declare.vue 1`] = `
"// declare.js
import { defineComponent, toDisplayString } from 'vue';
import { expectTypeOf } from 'expect-type';
import _export_sfc from '[NULL]/plugin-vue/export-helper';

var _sfc_main = /* @__PURE__ */ defineComponent({
	__name: "declare",
	props: { bar: {
		type: Boolean,
		required: false,
		default: true
	} },
	setup(__props, { expose: __expose }) {
		__expose();
		const props = __props;
		expectTypeOf(props).toEqualTypeOf();
		const __returned__ = { props };
		Object.defineProperty(__returned__, "__isScriptSetup", {
			enumerable: false,
			value: true
		});
		return __returned__;
	}
});

function _sfc_render(_ctx, _cache, $props, $setup, $data, $options) {
  return toDisplayString($setup.props);
}
var declare = /* @__PURE__ */ _export_sfc(_sfc_main, [["render", _sfc_render], __FILE__]);

export { declare as default };
"
`;

exports[`fixtures > tests/fixtures/empty.vue 1`] = `
"// empty.js
import { defineComponent } from 'vue';
import { expectTypeOf } from 'expect-type';
import _export_sfc from '[NULL]/plugin-vue/export-helper';

var _sfc_main = /* @__PURE__ */ defineComponent({
	__name: "empty",
	setup(__props, { expose: __expose }) {
		__expose();
		const props = __props;
		expectTypeOf().toEqualTypeOf();
		const __returned__ = {
			props,
			get expectTypeOf() {
				return expectTypeOf;
			}
		};
		Object.defineProperty(__returned__, "__isScriptSetup", {
			enumerable: false,
			value: true
		});
		return __returned__;
	}
});

var empty = /* @__PURE__ */ _export_sfc(_sfc_main, [__FILE__]);

export { empty as default };
"
`;

exports[`fixtures > tests/fixtures/empty-defaults.vue 1`] = `
"// empty-defaults.js
import { defineComponent } from 'vue';
import { expectTypeOf } from 'expect-type';
import _export_sfc from '[NULL]/plugin-vue/export-helper';

var _sfc_main = /* @__PURE__ */ defineComponent({
	__name: "empty-defaults",
	props: { foo: {
		type: String,
		required: false
	} },
	setup(__props, { expose: __expose }) {
		__expose();
		const props = __props;
		expectTypeOf().toEqualTypeOf();
		const __returned__ = {
			props,
			get expectTypeOf() {
				return expectTypeOf;
			}
		};
		Object.defineProperty(__returned__, "__isScriptSetup", {
			enumerable: false,
			value: true
		});
		return __returned__;
	}
});

var emptyDefaults = /* @__PURE__ */ _export_sfc(_sfc_main, [__FILE__]);

export { emptyDefaults as default };
"
`;

exports[`fixtures > tests/fixtures/tsx.vue 1`] = `
"// tsx.js
import { defineComponent } from 'vue';
import _export_sfc from '[NULL]/plugin-vue/export-helper';

var _sfc_main = /* @__PURE__ */ defineComponent({
	__name: "tsx",
	props: { foo: {
		type: String,
		required: false,
		default: "111"
	} },
	setup(__props, { expose: __expose }) {
		__expose();
		const __returned__ = {};
		Object.defineProperty(__returned__, "__isScriptSetup", {
			enumerable: false,
			value: true
		});
		return __returned__;
	}
});

var tsx = /* @__PURE__ */ _export_sfc(_sfc_main, [__FILE__]);

export { tsx as default };
"
`;
